### PR TITLE
Various improvements to SSGI

### PR DESCRIPTION
**Pull request by anisunity**
SSGI_Improve_2 (1)

This PR tries to improve the global quality of screen space global illumination. Here is a summary of the various changes.
Using the RTGI denoiser for SSGI.
Adding the fallback hierarchy control for SSGI (in case ray marching was not able to get a result)
Remove the fact that the ssgi signcal can be incomplete.
Move the temporal and diffuse denoisers to the non-raytracing part of the pipeline.
Normal history buffer requires when ssgi is on.
Increased the clamp value for SSGI

This PR should be solving these bugs.
https://fogbugz.unity3d.com/f/cases/1340851/
https://fogbugz.unity3d.com/f/cases/1327919/
https://fogbugz.unity3d.com/f/cases/1339297/

Testing status
Extensively tested on the template scene, results are really good.
Unfortunately due to the limiations of our test framework, still cannot add tests for SSGI

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/ScreenSpaceLighting/ScreenSpaceGlobalIllumination.compute [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
+#include "Packages/com.unity.render-pipelines.high-definition/Runtime/Lighting/LightEvaluation.hlsl"
+#include "Packages/com.unity.render-pipelines.high-definition/Runtime/Lighting/ScreenSpaceLighting/RayMarchingFallbackHierarchy.cs.hlsl"
 
 // Raytracing includes (should probably be in generic files)
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/ShaderVariablesRaytracing.hlsl"
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingSampling.hlsl"
 #include "Packages/com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RayTracingCommon.hlsl"
 
-// #pragma enable_d3d11_debug_symbols
+#pragma enable_d3d11_debug_symbols
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/ScreenSpaceLighting/ScreenSpaceGlobalIllumination.compute:22]: Remove this

**Comment by remi-chapelain** [comment]
Made a simple "inside" scene to test the different fallbacks solutions. The fallback on probe helps a lot eventhough the setup to make it look like what you want in terms of reflection probe placement could be tedious but it's a good solution while probe volume are experimental. 

Also roughly (only compared frame time for now and some variance) compared the cost of fallbacks and before the PR. 
My premature conclusion is that the cost is more or less the same (+/- 0.250ms). And that SSGI costs roughly 2 to 3ms as a whole. 

No SSGI										11.03 ms
SSGI Medium None fallback (Before PR)                        13.62 ms
SSGI Medium None fallback						13.72 ms
SSGI Medium Sky fallback						13.65 ms
SSGI Medium Reflection Probe Fallback			        13.75 ms
SSGI Medium Reflection Probe And Sky Fallback	        13.74 ms
SSGI Medium Probe Volume Fallback				13.48 ms
RTGI Perf mode in medium  		                                14.02 ms                                      
RTGI	Quality									24.13 ms

Ray Traced GI in medium perf mode cost only slightly more (roughly 4ms) and cost a lot in quality mode (12ms). 

https://user-images.githubusercontent.com/57442369/121927700-a6ccb780-cd3f-11eb-986d-43860c521f34.mp4

Here's the different tests images: 

**No SSGI**
S_Apartment_20210614_1705814_No SSGI

**SSGI Medium - No Fallback**
S_Apartment_20210614_1659322_SSGI Medium No Fallback

**SSGI Medium - Probe Volume Fallback**
S_Apartment_20210614_1733865_SSGI Probe Volume Fallback

**SSGI Medium - Reflection Probe Fallback**
S_Apartment_20210614_1700992_SSGI Medium Reflection Probe Fallback

**RTGI Quality**
S_Apartment_20210614_1706548_RTGI

**Path Tracer**
S_Apartment_20210614_1658808_Path Tracer

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/ScreenSpaceLighting/ScreenSpaceGlobalIllumination.compute (hunk 1 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/ScreenSpaceLighting/ScreenSpaceGlobalIllumination.compute]: done

**Comment by anisunity** [comment]
Fixed the APV issue reported by @remi-chapelain

**Review by sebastienlagarde — APPROVED** [review]

**Comment by remi-chapelain** [comment]
We also need to append the documentation about the light probe > reflection probe fallback hierarchy :)

**Comment by sebastienlagarde** [comment]
there is various 
<html>
<body>


Test Name: HDRP_GraphicTestRunner.002-HDMaterialsTest Fixture: HDRP_RuntimeTestsTest Status: FailureTest Execution Time: 00:00:26.192Message: Copy to ClipboardUnhandled log message: '[Error] Kernel 'ValidateHistory' not found'. Use UnityEngine.TestTools.LogAssert.Expect
--



</body>
</html>  error in Yamato

**Comment by anisunity** [comment]
Probably a resource thing, i'll check that

**Review by remi-chapelain — CHANGES_REQUESTED** [review]
Finished testing. 
Tested on the template and on various projects to see if the position of the reflections probe needed to be different to be able to have a correct results with SSGI. And it's fine in most projects. Probes position can still be adjusted to change SSGI results for artistic reasons but it's a case by case. 

**Comparison between no fallback and the new reflection probe fallback**
https://user-images.githubusercontent.com/57442369/124471533-43153780-dd9d-11eb-8311-19ef81b24cfb.mp4

**Comparison between SSGI and no SSGI**
https://user-images.githubusercontent.com/57442369/124471820-a010ed80-dd9d-11eb-9378-37b45c8cb72f.mp4

Only things I'd like to see changed before fully approving are : 
- Documentation to be updated about the new fallbacks and the hierarchy of it (probe volume first, then reflection probes).
- High quality settings denoiser radius needs to be increased to 0.5 (instead of 0.25). In some cases, you can still see "blobs" of lights with .25 radius which looks worse than medium, especially when you are moving.

**Comment by anisunity** [comment]
I pushed a fix that should fix all failing tests

**Review by remi-chapelain — APPROVED** [review]
🟢

**Comment by anisunity** [comment]
Merged master and relaunched the tests

**Pull request merged**